live/live-root/usr/bin/live-self-update

```diff
@@ -23,56 +48,140 @@ run_command() {
 # Prepare the system root to be used by scripts before switching to it.
 prepare_chroot() {
   if [ -n "$NEWROOT" ]; then
-    echo "<5>Preparing for the self update..."
+    milestone "Preparing for the self update..."
 
-    # make local devices available to "agama download"
+    # make system directories active also in the chroot directory
     mount -o bind /run "$NEWROOT"/run
     mount -o bind /sys "$NEWROOT"/sys
     mount -o bind /proc "$NEWROOT"/proc
     mount -o bind /dev "$NEWROOT"/dev
-    cp /run/NetworkManager/resolv.conf "$NEWROOT"/etc/resolv.conf
+    [ -f /run/NetworkManager/resolv.conf ] && cp /run/NetworkManager/resolv.conf "$NEWROOT"/etc/resolv.conf
   fi
 }
 
 # Revert the changes done by prepare_chroot() and restore the system to the
-# original state
+# original state, called from bash "trap" handler
 cleanup_chroot() {
   if [ "$?" != "0" ]; then
-    echo "<3>Error: Self update failed!"
+    error "Error: Self update failed!"
   else
-    echo "<5>Self update successfully finished"
+    milestone "Self update successfully finished"
   fi
 
   if [ -n "$NEWROOT" ]; then
-    rm "$NEWROOT"/etc/resolv.conf
-    umount "$NEWROOT"/run
-    umount "$NEWROOT"/sys
-    umount "$NEWROOT"/proc
-    umount "$NEWROOT"/dev
+    rm -f "$NEWROOT"/etc/resolv.conf
+    # do not fail if not mounted
+    findmnt "$NEWROOT"/run > /dev/null && umount "$NEWROOT"/run
+    findmnt "$NEWROOT"/sys > /dev/null && umount "$NEWROOT"/sys
+    findmnt "$NEWROOT"/proc > /dev/null && umount "$NEWROOT"/proc
+    findmnt "$NEWROOT"/dev > /dev/null && umount "$NEWROOT"/dev
   fi
 }
 
-# Applies all the updates
-#
-# Reads the URL of the updates from $AGAMA_DUD_INFO and process each one.
-install_updates() {
-  prepare_chroot
+# read the data from /etc/os-release file (source it)
+read_os_release() {
+  # get the product data
+  if [ -f "$OSR" ]; then
+    . "$OSR"
+  else
+    error "File $OSR is missing!"
+    exit 1
+  fi
+}
 
-  index=0
-  self_update_repos=()
+# ask the registration server (SCC or RMT) for the self-update URL
+query_reg_server() {
+  local server
+  # custom server (RMT)
+  if [ -f "$SERVER_FILE" ]; then
+    server=$(cat "$SERVER_FILE")
+  # default server (SCC)
+  elif  [ -f "$CONFIG_DEFAULT_REG_SERVER_FILE" ]; then
+    server=$(cat "$CONFIG_DEFAULT_REG_SERVER_FILE")
+  fi
+
+  if [ -n "$server" ]; then
+    read_os_release
+
+    local query
+    query="$server/connect/repositories/installer?identifier=$NAME&version=$VERSION_ID&arch=$ARCH"
+    info "Running query $query...."
 
-  echo "<5>Adding self update repositories..."
-  # add the self-update repositories  
+    curl --silent --globoff --location --retry 3 --connect-timeout 20 --max-time 10 --fail --show-error --output "$RESPONSE_FILE" -- "$query"
```

Comment:
did you considered use the **fetch_url** function from the **dracut url-lib** module?

https://github.com/dracutdevs/dracut/blob/master/modules.d/45url-lib/url-lib.sh#L59

Reply:
Yes, I already tried using `dracut-lib.sh`. But the problem is that it works only when running from dracut. It does not work after switching to the root image. And I'd like to make it work also in the root image as that makes testing and debugging easier. And more importantly it allows users manually restart the self-update if it for whatever reason failed in initrd.